code/game/machinery/vending.dm

```diff
@@ -766,6 +718,99 @@ var/global/num_vending_terminals = 1
 		return 1
 	return 0
 
+/obj/machinery/vending/tgui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Vending", name)
+		ui.set_autoupdate(FALSE)
+		ui.open()
+
+/obj/machinery/vending/ui_static_data(mob/user)
+	var/list/data = list()
+	data["ad"] += pick(product_ads)
+	data["has_premium"] = coin_records.len > 0
+	data["bypass"] = isAdminGhost(user)
+	data["silicon"] = istype(user, /mob/living/silicon)
+	if (data["silicon"] && istype(user, /mob/living/silicon/robot))
+		var/mob/living/silicon/robot/R = user
+		if (HAS_MODULE_QUIRK(R, MODULE_CAN_BUY))
+			data["bypass"] = TRUE // act like normal if the module allows it
+	return data
+
+/obj/machinery/vending/ui_data(mob/user)
+	var/list/data = list()
+	if (currently_vending != null)
+		data["currently_vending"] = list(
+			name = currently_vending.product_name,
+			price = currently_vending.price,
+			ref = ref(currently_vending)
+		)
+	else
+		data["currently_vending"] = null
+		data["edit_mode"] = edit_mode
+		data["machine_name"] = name
+		data["vend_ready"] = vend_ready
+		data["contraband"] = extended_inventory
+		data["coin"] = coin != null
+		data["categories"] = list()
+		for (var/datum/data/vending_product/P in get_all_records())
+			var/list/product = list(
+				name = P.product_name,
+				amount = P.amount,
+				price = P.price,
+				category = P.category,
+				icon = P.mini_icon,
```

Comment:
Sending the base64 for all the icons with *every* interaction is the only thing in here that I find suspicious. How does a vending machine with a large inventory such as the Autodrobe perform with this?